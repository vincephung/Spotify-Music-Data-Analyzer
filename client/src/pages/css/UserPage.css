
.profileCard{
    margin: auto;
    margin-top: 100px;
    display: flex;
    justify-content: center;
    position: relative;
    width: 75%;
    font-family: Rowdies;
    background-color: #5E5C5C;
    border: 2px solid;
    padding: 0.5em 0.5em;
    box-shadow: 3px 2px 0px 0px , 3px 3px 0px 0px, 4px 4px 0px 0px, 6px 6px 0px 0px, 8px 8px 0px 0px;
}

.profileText{
    color: rgb(209, 209, 209);
}


.profilePicture{
    width: 25%;
}

.profilePictureDiv{
    padding: 20px;
    display:flex;
    justify-content: center;
}


.profileBtns{
    background-color: #00B2FF;
    font-family: Rowdies;
    font-size: 14px;
    color: #000;
    cursor: pointer;
    border: 3px solid;
    padding: 0.25em 0.5em;
    box-shadow: 1px 1px 0px 0px, 2px 2px 0px 0px, 3px 3px 0px 0px, 4px 4px 0px 0px, 5px 5px 0px 0px;
    position: relative;
    touch-action: manipulation;

}


/** For Edit User Page Form **/
.editUserPageContainer{
    display: flex;
    flex-direction: column; 
    align-items: center;
}

.userForm{
    border-style: solid;
    border-color: black;
    border-width: 2px;
    border-radius: 10%;
    width: 50%;
    padding: 40px;
}
