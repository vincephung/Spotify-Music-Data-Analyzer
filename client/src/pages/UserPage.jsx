import './css/UserPage.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import { Button, Card, Image, Stack, Link } from 'react-bootstrap'
import toast, { Toaster } from 'react-hot-toast';
import React from 'react';
import { checkTokens, logOut } from '../manageTokens'
import { useEffect, useState } from 'react';
import defaultImage from './img/profile.png'

const UserPage = () => {
    useEffect(checkTokens, [])

    const [ profile, setProfile ] = useState(() => {
        return JSON.parse(
            window.sessionStorage.getItem('userProfile')
        )
    });
    const [ premium, setPremium ] = useState(profile.isPremium);

    // For display on profile
    let premiumStatus = "Standard";
    if (premium){
        premiumStatus = "Premium";
    }
    let shareLink = "";
    if (profile) {
        shareLink = `${window.location.origin}/premium/sharing/${profile.generatedShareLink}`;
    }

    return (
        <>{
            profile && (
                <Card className="profileCard text-center">
                    <div className="profilePictureDiv">
                        <Image className="profilePicture" roundedCircle thumbnail src={
                            profile.profile_img || defaultImage
                        }></Image>
                    </div>
                    <div className="profileText">
                        <Card.Title>{profile.display_name}</Card.Title>
                        <Card.Text></Card.Text>
                        <Card.Text>Username: {profile.id}</Card.Text>
                        <Card.Text>Email: {profile.email}</Card.Text> 
                        <Card.Text>Location: {profile.country}</Card.Text>
                        <Card.Text>Plan: {premiumStatus}</Card.Text>
                        { profile.generatedShareLink && premium && <Card.Text>
                            Share Link: <a href={shareLink} className="text-decoration-none">
                                {shareLink}
                            </a>
                        </Card.Text> }
                    </div>
                    <Card.Footer style={{padding: '10px'}}>
                        <Stack direction="horizontal" className="justify-content-md-center" gap={3}>
                            <Button 
                                className="profileBtns" variant ="success" 
                                onClick={() => {
                                    fetch(`/api/togglePremium?email=${profile.email}`)
                                    .then(result => {
                                        result.json(result).then(result => {
                                            if (result["error"]) {
                                                // Report error
                                                console.log("Error toggling premium:")
                                                console.log(result["error"])
                                                return;
                                            }

                                            console.log(`Successfully toggled premium status (${result['success']})`);

                                            // Update the profile object
                                            profile.isPremium = result["success"];
                                            window.sessionStorage.setItem("userProfile", JSON.stringify(profile));
                                        
                                            setPremium(profile.isPremium)
                                        });
                                    });
                                }
                            }> 
                                { premium ? 'Quit' : 'Join'} Premium 
                            </Button>
                            <Button 
                                className="profileBtns" variant ="danger" 
                                onClick={() => {

                                    const access_token = localStorage.getItem('access_token');

                                    fetch(`/api/refresh?email=${profile.email}&access_token=${access_token}`)
                                    .then(result => {
                                        result.json(result).then(result => {
                                            if (result["error"]) {
                                                // Report error
                                                console.log("Error refreshing data:")
                                                console.log(result["error"])
                                                return;
                                            }

                                            // Log the success
                                            console.log(`Successfully refreshed data!`);
                                        });
                                    });
                                }
                            }> Refresh Data</Button>
                            <Button 
                                className="profileBtns" 
                                variant ="danger" 
                                disabled={!premium}
                                onClick={() => {
                                    // Fetch to the generate share link endpoint
                                    fetch(`/api/user/generateShareLink/${profile.email}`)
                                    .then((result) => {
                                        // Convert the returned result to JSON
                                        result.json().then((result) => {

                                            // Report the error, if it exists
                                            if (result["error"]) {
                                                console.log(`There was an error generating share link`)
                                                console.log(result["error"])
                                                return;
                                            }

                                            // Otherwise, continue
                                            console.log(`Successfully retrieved share link: ${result["success"]}`)
                                            
                                            // Create a new profile object with all the old fields, and add the generatedShareLink 
                                            //      field as well
                                            const newProfile = { 
                                                ...profile, 
                                                generatedShareLink: result["success"]
                                            }

                                            // Write the link to the clipboard, and report this with a toast
                                            navigator.clipboard.writeText(`${window.location.origin}/premium/sharing/${result["success"]}`)
                                            toast.success("Copied share link to clipboard")

                                            // Add the profile to session storage, and update the profile state 
                                            sessionStorage.setItem("userProfile", JSON.stringify(newProfile))
                                            setProfile(newProfile)

                                        })
                                        .catch((err) => {
                                            console.log(`There was an error generating share link`)
                                            console.log(err)
                                        })
                                    })
                                    .catch((err) => {
                                        console.log(`There was an error generating share link`)
                                        console.log(err)
                                    })


                                }}>Generate Share Link</Button>
                            <Button className="profileBtns" variant ="danger" onClick={logOut}>Log out</Button>
                        </Stack>
                    </Card.Footer>
                </Card> 
            )}
            <div><Toaster
                position="top-center"
                reverseOrder={false}
            /></div>
        </>
        
    );
}

export default UserPage