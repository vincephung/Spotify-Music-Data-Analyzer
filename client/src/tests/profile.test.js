import { render, unmountComponentAtNode  } from 'react-dom';
import { act } from "react-dom/test-utils";
import UserPage from '../pages/UserPage';

let container = null;

let profile = {
    country: "US",
    display_name: "statify",
    email: "statifytest@gmail.com",
    generatedShareLink: "http://localhost:8400/premium/sharing/VbXqvHkWfxvMSNor",
    id: "31oadqabwvmlmape6n5uyccvqgcm",
    isPremium: false,
    product: "free",
    profile_img: null,
}

beforeEach(() => {
sessionStorage.setItem('userProfile', JSON.stringify(profile));
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});


test("Get profile page", async () => {
    act(() => {
        render(<UserPage/>, container);
    });
    expect(container).toMatchSnapshot()
})