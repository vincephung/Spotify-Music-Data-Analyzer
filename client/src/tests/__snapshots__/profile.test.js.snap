// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get profile page 1`] = `
<div>
  <div
    class="profileCard text-center card"
  >
    <div
      class="profilePictureDiv"
    >
      <img
        class="profilePicture rounded-circle img-thumbnail"
        src="profile.png"
      />
    </div>
    <div
      class="profileText"
    >
      <div
        class="card-title h5"
      >
        statify
      </div>
      <p
        class="card-text"
      />
      <p
        class="card-text"
      >
        Username: 
        31oadqabwvmlmape6n5uyccvqgcm
      </p>
      <p
        class="card-text"
      >
        Email: 
        statifytest@gmail.com
      </p>
      <p
        class="card-text"
      >
        Location: 
        US
      </p>
      <p
        class="card-text"
      >
        Plan: 
        Standard
      </p>
    </div>
    <div
      class="card-footer"
      style="padding: 10px;"
    >
      <div
        class="justify-content-md-center hstack gap-3"
      >
        <button
          class="profileBtns btn btn-success"
          type="button"
        >
          Join
           Premium
        </button>
        <button
          class="profileBtns btn btn-danger"
          type="button"
        >
           Refresh Data
        </button>
        <button
          class="profileBtns btn btn-danger"
          disabled=""
          type="button"
        >
          Generate Share Link
        </button>
        <button
          class="profileBtns btn btn-danger"
          type="button"
        >
          Log out
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
    />
  </div>
</div>
`;
